src/main/java/io/cdap/plugin/snowflake/common/client/SnowflakeAccessor.java

```diff
@@ -0,0 +1,178 @@
+  private static final String APPLICATION_NAME = "CDAP";
+  private static final int LIMIT_ROWS = 5;
+
+  private final BaseSnowflakeConfig config;
+  protected final SnowflakeBasicDataSource dataSource;
+
+  public SnowflakeAccessor(BaseSnowflakeConfig config) {
+    this.config = config;
+    this.dataSource = new SnowflakeBasicDataSource();
+    initDataSource(dataSource, config);
+  }
+
+  public void runSQL(String query) throws IOException {
+    try (Connection connection = dataSource.getConnection();
+         PreparedStatement populateStmt = connection.prepareStatement(query);) {
+      populateStmt.execute();
+    } catch (SQLException e) {
+      throw new IOException(String.format("Statement '%s' failed.", query), e);
+    }
+
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(SnowflakeAccessor.class);
+
+  /**
+   * Returns field descriptors for specified import query.
+   *
+   * @return List of field descriptors.
+   * @throws IOException thrown if there are any issue with the I/O operations.
+   */
+  public List<SnowflakeFieldDescriptor> describeQuery(String query) throws IOException {
+    String importQuery = QueryUtil.limitQuery(query, LIMIT_ROWS);
+    List<SnowflakeFieldDescriptor> fieldDescriptors = new ArrayList<>();
+
+    try (Connection connection = dataSource.getConnection();
+         PreparedStatement preparedStatement = connection.prepareStatement(importQuery);
+         ResultSet resultSet = preparedStatement.executeQuery()) {
+      ResultSetMetaData metaData = resultSet.getMetaData();
+      int columnCount = metaData.getColumnCount();
+      for (int i = 1; i <= columnCount; i++) {
+        String name = metaData.getColumnName(i);
+        int type = metaData.getColumnType(i);
+        boolean nullable = metaData.isNullable(i) == ResultSetMetaData.columnNullable;
+        fieldDescriptors.add(new SnowflakeFieldDescriptor(name, type, nullable));
+      }
+    } catch (SQLException e) {
+      throw new IOException(e);
+    }
+    return fieldDescriptors;
+  }
+
+  private void initDataSource(SnowflakeBasicDataSource dataSource, BaseSnowflakeConfig config) {
+    dataSource.setDatabaseName(config.getDatabase());
+    dataSource.setSchema(config.getSchemaName());
+    dataSource.setUrl(String.format("jdbc:snowflake://%s.snowflakecomputing.com", config.getAccountName()));
+
+    String warehouse = config.getWarehouse();
+    if (!Strings.isNullOrEmpty(warehouse)) {
+      dataSource.setWarehouse(warehouse);
+    }
+
+    String role = config.getRole();
+    if (!Strings.isNullOrEmpty(role)) {
+      dataSource.setRole(role);
+    }
+
+    if (config.getOauth2Enabled()) {
+      String accessToken = OAuthUtil.getAccessTokenByRefreshToken(HttpClients.createDefault(), config);
+      dataSource.setOauthToken(accessToken);
+    } else if (config.getKeyPairEnabled()) {
+      dataSource.setUser(config.getUsername());
+
+      String privateKeyPath = writeTextToTmpFile(config.getPrivateKey());
```

Comment:
where will this private key get serialized?

Reply:
also why does this file need to be serialized?

Reply:
Jdbc snowflake API excepts parameter which is file path.
```dataSource.setPrivateKeyFile(privateKeyPath, config.getPassphrase());```
dataSource is an instance of class from snowflake jdbc API.